clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -2141,15 +2141,13 @@ Address CodeGenFunction::EmitIntelFPGAFieldAnnotations(SourceLocation Location,
                                                        StringRef AnnotStr) {
   llvm::Value *V = Addr.getPointer();
   llvm::Type *VTy = V->getType();
+  unsigned AS = VTy->getPointerAddressSpace();
+  llvm::Type *Int8VPtrTy = llvm::Type::getInt8PtrTy(CGM.getLLVMContext(), AS);
   llvm::Function *F =
-      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, CGM.Int8PtrTy);
-  // FIXME Always emit the cast inst so we can differentiate between
-  // annotation on the first field of a struct and annotation on the struct
-  // itself.
-  if (VTy != CGM.Int8PtrTy)
-    V = Builder.CreateBitCast(V, CGM.Int8PtrTy);
+      CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, Int8VPtrTy);
+  V = Builder.CreatePointerBitCastOrAddrSpaceCast(V, Int8VPtrTy);
```

Comment:
If `V->getType()` and `Int8VPtrTy` have the same address space, then you don't need `CreatePointerBitCastOrAddrSpaceCast`, and `CreateBitCast` should suffice. You also don't need a bitcast if `V->getType()` is a pointer to an integer (of any width).

Reply:
Fixed